File: Octave/descenso_coordinado.m
## Copyright (C) 2019 mherr
## 
## This program is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
## 
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see
## <https://www.gnu.org/licenses/>.

## -*- texinfo -*- 
## @deftypefn {} {@var{retval} =} descenso_coordinado (@var{input1}, @var{input2})
##
## @seealso{}
## @end deftypefn

## Author: mherr <mherr@MSI>
## Created: 2019-08-27

## Metodo iterativo del Descenso Coordinado para optimizacion en varias
## variables.
## Entradas: valores iniciales, variables de la funcion, tolerancia minima
## del resultado y funcion a evaluar en string.
## Salidas:Valor aproximado y numero de iteraciones
function [x0, iteration] = descenso_coordinado(x0, str_variables, tol, funcion)
    warning('off');
    n = length(str_variables);
    f = str2func(funcion);
    error = tol;
    iteration = 0;
    for i = 1:n
        variables(i) = sym(str_variables(i));
    endfor
    
    function x0 = replace(x0, indice, variables, f)
        temp = num2cell(x0);
        temp(indice) = variables(indice);
        df = diff(f(temp{:}));
        soluciones = solve(df);
        m = length(soluciones);
        for i = 1:m
            temp(indice) = double(soluciones(i));
            if f(temp{:}) < f(num2cell(x0){:})
            x0 = cell2mat(temp);
            endif
        endfor
    endfunction
    
    
    while error >= tol
        prev = x0;
        for j = 1:n
            x0 = replace(x0, j, variables, f);
        endfor
        iteration += 1;
        error = norm(x0 - prev);
    endwhile
endfunction

File: Python/descenso_coordinado.py
from sympy import *
import math
import numpy as np


"""
Metodo iterativo del Descenso Coordinado para optimizacion
en varias variables.
Entradas: valores iniciales, variables de la funcion, tolerancia minima
del resultado y funcion a evaluar en string.
Salidas:Valor aproximado y numero de iteraciones
"""
def descenso_coordinado(x0, variables, tol, funcion):
    n = len(variables)
    iteracion = 0
    for i in range(n):
        variables[i] = Symbol(variables[i])
    funcion = sympify(funcion)
    f = lambdify(variables, funcion, "numpy")
    error = tol
    while error >= tol:
        prev = x0.copy()
        for j in range(n):
            x0 = replace(x0, j, variables[j], f)
        iteracion += 1
        difference = np.array(np.subtract(x0, prev), dtype=np.float64)
        error = np.linalg.norm(difference)

    return x0, iteracion

"""
Metodo auxiliar del Descenso Coordinado para calcular el minimo en una variable
y reemplazarlo
Entradas: valores iniciales, indice de la variable a optimizar,
variables de la funcion, funcion
Salidas: iteracion ya reemplazada
"""
def replace(x0, indice, variable, f):
    temp = x0.copy()
    temp[indice] = variable
    df = lambdify(variable, diff(f(*temp), variable), "numpy")
    soluciones = solve(df(variable))
    for solucion in soluciones:
        temp[indice] = solucion
        if(f(*temp) < f(*x0)):
            x0 = temp
    return x0
